Add `ddev clean` command, fixes #2332

mikebarkas opened the pull request:
## The Problem/Issue/Bug:
Provide a new clean command

## How this PR Solves The Problem:
This creates a new command

## Manual Testing Instructions:
Run `ddev clean projectname` and verify the snapshots have been removed

## Automated Testing Overview:
No tests have been created. I would like to create the tests.

## Related Issue Link(s):
#2332 

## Release/Deployment notes:
<!-- Does this affect anything else, or are there ramifications for other code? Does anything have to be done on deployment? -->
No



<a href="https://gitpod.io/#https://github.com/drud/ddev/pull/3683"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -0,0 +1,99 @@
+	Long: `Stops all running projects and then removes downloads and snapshots
+for the selected projects. Then clean will remove ddev images.
+
+Warning - This command will permanently delete your snapshots and you will lose
+your database data in these snapshots.
```
```suggestion
Warning - This command will permanently delete your snapshots for the named project[s].
```

mikebarkas replied:
Updated, thanks.

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -0,0 +1,99 @@
+			os.Exit(1)
+		}
+
+		if util.Confirm("Confirm removal of the listed items?") {
```
How about using util.Prompt() here instead, with default 'N' ?

mikebarkas replied:
Like this?
```
    confirm := util.Prompt("Confirm removal of the listed items? Y/N", "N")
    if confirm == "Y"  {
```

rfay replied:
Lower-case n will work right IIRC, lower-case `y` won't here. But don't ask me... try it out! Step through it, improve it :)

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -0,0 +1,99 @@
+		}
+
+		if util.Confirm("Confirm removal of the listed items?") {
+			ddevDir := globalconfig.GetGlobalDdevDir()
```
Let's use a variable that explains what it is here, maybe globalDdevDir. I always hate when I confuse the two dirs.

mikebarkas replied:
Good idea, thanks. I changed it.

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -0,0 +1,99 @@
+			ddevDir := globalconfig.GetGlobalDdevDir()
+			_ = os.RemoveAll(filepath.Join(ddevDir, "testcache"))
+			_ = os.RemoveAll(filepath.Join(ddevDir, "bin"))
+			_ = os.RemoveAll(filepath.Join(ddevDir, "downloads"))
```
I don't think the global ddev dir has ever had a downloads folder.

mikebarkas replied:
Yeah, I confused this with the .downloads at the project level.
I will fix.

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -0,0 +1,99 @@
+			output.UserOut.Print("Deleting snapshots and downloads for selected projects...")
+			for _, project := range projects {
+				// Delete snapshots and downloads for each project
+				_ = os.RemoveAll(filepath.Join(project.AppRoot, ".ddev", "downloads"))
```
Easier usage for these is app.GetConfigPath("downloads")

But I think the thing we want to delete is ".downloads", with the dot on the front.

mikebarkas replied:
Ah.. that method is better. I renamed to ".downloads" too. Thanks.

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -0,0 +1,99 @@
+				_ = os.RemoveAll(filepath.Join(project.AppRoot, ".ddev", "db_snapshots"))
+			}
+
+			output.UserOut.Print("Deleting images that ddev created...")
```
```suggestion
			output.UserOut.Print("Deleting docker images that ddev created...")
```

People are so confused about images/containers/docker that we can help them a little here.

mikebarkas replied:
I added docker

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -0,0 +1,99 @@
+
+			output.UserOut.Print("Deleting images that ddev created...")
+			client := dockerutil.GetDockerClient()
+			images, err := client.ListImages(docker.ListImagesOptions{
```
Let's promote the code you wrote for deleting all images to a function and not duplicate the code.

mikebarkas replied:
I will move this to the function.

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -0,0 +1,99 @@
+			if err != nil {
+				util.Failed("Failed to list images: %v", err)
+			}
+			err = deleteDdevImages(images)
```
Oh, I see you promoted, but didn't promote all of it. Shouldn't you be able to promote the whole thing? With a deleteAll arg?

mikebarkas replied:
I will move this to the function.

mikebarkas replied:
I think it makes sense to have `deleteDdevImages()` get its images and have a deleteAll argument.

I can then move the delete all logic to be first in the command.

mikebarkas replied:
I implemented this.

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -0,0 +1,99 @@
+
+func init() {
+	CleanCmd.Flags().BoolP("all", "a", false, "Clean all ddev projects")
+	CleanCmd.Flags().Bool("dry-run", false, "Run the clean command without deleting")
```
Thanks for trying out .Bool() - I just hate all those single-purpose variables everywhere.

rfay commented on cmd/ddev/cmd/delete-images.go:
```diff
@@ -130,3 +122,21 @@ func init() {
 }
+
+// deleteDdevImages removes Docker images prefixed with ddev-
+func deleteDdevImages(images []docker.APIImages) error {
```
As above, I think this can be done more completely by making this function find the images itself. Then the arg would be `deleteAll bool`. Without deleteAll it would just do old images.

mikebarkas replied:
How do we define "old" images?

rfay replied:
The "old" images are the ones that don't go with the current version. That's what you did in the delete images work - looking to see what matched current and what didn't. And the --all was all of them regardless of match.

rfay reviewed (changes requested):
Making progress, thanks so much for this!

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -62,28 +64,21 @@ Additional commands that can help clean up resources:
 			output.UserOut.Print("Deleting snapshots and downloads for selected projects...")
 			for _, project := range projects {
 				// Delete snapshots and downloads for each project
-				_ = os.RemoveAll(filepath.Join(project.AppRoot, ".ddev", "downloads"))
```
Let's check the error and `util.Warn()` about it, but not fail. If we found it and can't delete it, there's something going on.

mikebarkas replied:
I will add the error check.

rfay commented:
You'll want to rebase this when you get back to it, thanks.

mikebarkas commented:
@rfay 

- I refactored the delete images function to be reusable by other commands.
- I removed the flag variable placeholders for delete images and used .BoolP
- Created a vscode debug launcher for delete images command
- Compiled and tested locally deleting images with and without --all
- Rebased local branch and pushed

This is ready for review when the builds and tests finish.

rfay commented:
It seems like the CircleCI failure is a problem with Pantheon, so can be ignored. It's failing there on other tests as well.

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -0,0 +1,101 @@
+		dryRun, _ := cmd.Flags().GetBool("dry-run")
+		if dryRun {
+			util.Warning("Dry run terminated without removing items")
+			os.Exit(1)
```
No biggie, but I think this should just be a `return`, which will exit success rather than error.

If it weren't to be a success (which seems ok with dry run) then I would have terminated it with utiil.Failed() which does an exit.

mikebarkas replied:
ok, I agree.

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -0,0 +1,101 @@
+		}
+
+		confirm := util.Prompt("Are you sure you want to continue? y/n", "n")
+		if strings.ToLower(confirm) == "y" {
```
I think it would be more natural to check for *not* "y" and return if not. Then the rest could continue without being in an "if" stanza.

mikebarkas replied:
Happy path.. is better.

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -0,0 +1,101 @@
+			output.UserOut.Print("Deleting snapshots and downloads for selected projects...")
+			for _, project := range projects {
+				// Delete snapshots and downloads for each project
+				err = os.RemoveAll(project.GetConfigPath(".downloads"))
```
I didn't know that RemoveAll() returns nil if the path doesn't exist, so this works out fine.

rfay replied:
This would be cleaner with a for loop of a slice of strings, since the action is the same everywhere. No change required.

rfay commented on cmd/ddev/cmd/clean.go:
```diff
@@ -0,0 +1,101 @@
+			}
+
+			output.UserOut.Print("Deleting Docker images that ddev created...")
+			err = deleteDdevImages(cleanAll)
```
cleanAll was set based on "all projects" here, but your function is deciding between "all DDEV images" and "current DDEV images", which is a different decision.

IMO you can just use `true` here.

mikebarkas replied:
I had true there and then changed it. haha.. I will change it back, I agree.

rfay reviewed (changes requested):
Looks like almost done, thanks for all the good work. I think there are a couple of comments here that are actionable.

rfay reviewed (approved):
Looks good! I tested on macOS and Linux, blew away everything a couple of times. 

Please take another read-through looking especially for mistakes that might cause a deletion-of-wrong-thing disaster, and let me know when you're ready for me to pull it.

Congrats!

mikebarkas commented:
Looks good to me, I think it can be merged. Thanks for your feedback and guidance, it is appreciated.
